### PR TITLE
CMake refactor stage1

### PR DESCRIPTION
**Pull request by friendlyanon**
This PR is the first stage of refactoring the CMake build files of simdjson to remove spaghetti code, increase ease of use for consumers and developers of simdjons, and to make @lemire less angry at CMake in commit messages :)

Developer mode
---

This PR separates consumer and developer targets using a new option: `SIMDJSON_DEVELOPER_MODE`

This option is set to `OFF` by default (forced to be `ON` in CI) and enabling it makes tests, examples, benchmarks and others available. With this setting `ON`, it's as if nothing happened.

Consumer side fixes
---

simdjson now uses builtin CMake variable `BUILD_SHARED_LIBS` to switch between static or shared. Some places were highlighted with `FIXME`s that will be later changed wrt building shared libraries. I have an [educational repository](https://github.com/friendlyanon/cxx-static-shared-example) showcasing what and how should be done in a CMake project to enable building a cross-platform library as either static or shared. You can find more information there on what has to be changed exactly.

Install rules are now properly componentized into separate runtime and developer components (think `libA` and `libA-dev` packages from Linux distros). For this reason, minimum CMake requirement was raised to 3.14, so when omitting the `DESTINATION` arguments to the blocks in `install(TARGETS ...)`, then they will be set to the appropriate values from `GNUInstallDirs`. This is not a hard requirement, it just makes that command less wordy.

The next stage
---

In later PRs, I am going to work on the developer mode targets to clean things up and make the CMake scripts pretty.

### CODE REVIEW

**Comment by friendlyanon** [comment]
I couldn't get `ubuntu18-checkperf.yml` and the failing CircleCI jobs to fail locally, so this is ready for review.

File: benchmark/checkperf.cmake [hunk 1]
```diff
@@ -60,7 +60,7 @@ if (Git_FOUND AND (GIT_VERSION_STRING VERSION_GREATER  "2.1.4") AND (NOT CMAKE_G
   add_custom_command(
     OUTPUT ${SIMDJSON_CHECKPERF_DIR}/build/cmake_install.cmake # We make many things but this seems the most cross-platform one we can depend on
     COMMAND
-    ${CMAKE_COMMAND} -E env CXX=${CMAKE_CXX_COMPILER} CC=${CMAKE_C_COMPILER}
+    ${CMAKE_COMMAND} -E env CXX=${CMAKE_CXX_COMPILER} CC=${CMAKE_C_COMPILER} CI=$ENV{CI}
```
Comment (jkeiser) [inline benchmark/checkperf.cmake:63]: I'm super puzzled: I see mention of this in the docs but can't find where we do if (CI) in our cmake code, either in this PR or in my tree.

File: cmake/developer-options.cmake [hunk 2]
```diff
@@ -0,0 +1,171 @@
+#
```
Comment (jkeiser) [inline cmake/developer-options.cmake:1]: simdjson-options was getting pretty big, thanks for splitting it up :)

File: cmake/simdjson-config.cmake.in [hunk 3]
```diff
@@ -1,9 +1,6 @@
-@PACKAGE_INIT@
-
 include(CMakeFindDependencyMacro)
-if(@SIMDJSON_ENABLE_THREADS@)
-    set(CMAKE_THREAD_PREFER_PTHREAD TRUE)
```
Comment (jkeiser) [inline cmake/simdjson-config.cmake.in:5]: Why isn't this needed anymore?

File: cmake/simdjson-props.cmake [hunk 4]
```diff
@@ -0,0 +1,47 @@
+#
```
Comment (jkeiser) [inline cmake/simdjson-props.cmake:1]: Mind a brief comment on the intent of this? I'm having trouble parsing the intent from the code.

File: cmake/simdjson-user-cmakecache.cmake [hunk 5]
```diff
@@ -1,24 +1,19 @@
 #
-# ${SIMDJSON_USER_CMAKECACHE} contains the *user-specified* simdjson options so you can call cmake on
-# another branch or repository with the same options.
+# ${USER_CMAKECACHE} contains the *user-specified* simdjson options so you can
```
Comment (jkeiser) [inline cmake/simdjson-user-cmakecache.cmake:2]: What's the thought behind removing SIMDJSON_? Is this a standard variable?

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -1,161 +0,0 @@
-#
```
Comment (jkeiser) [inline src/CMakeLists.txt:1]: I'm really surprised at removing this! Isn't cmake idiom usually to have a file per directory? (I honestly don't know, I still feel like I barely understand the whole convoluted mess that is cmake :)

File: tests/CMakeLists.txt [hunk 7]
```diff
@@ -9,11 +9,19 @@ add_subdirectory(ondemand)
 #
 if(NOT SIMDJSON_LEGACY_VISUAL_STUDIO AND NOT SIMDJSON_WINDOWS_DLL)
   add_cpp_test(numberparsingcheck LABELS acceptance per_implementation)
-  target_link_libraries(numberparsingcheck simdjson-include-source simdjson-windows-headers)
+  target_link_libraries(numberparsingcheck simdjson simdjson-windows-headers)
```
Comment (jkeiser) [inline tests/CMakeLists.txt:12]: I'm curious how this works ... I would have thought using -Lsimdjson and simultaneously #including simdjson.cpp would cause duplicate link errors?

File: tools/CMakeLists.txt [hunk 8]
```diff
@@ -1,10 +1,13 @@
 if(TARGET cxxopts) # we only build the tools if cxxopts is available
   message(STATUS "We have cxxopts as a dependency and we are buiding the tools (e.g., json2json).")
-  link_libraries(simdjson simdjson-internal-flags simdjson-windows-headers cxxopts)
-  add_executable(json2json json2json.cpp)
-  add_executable(jsonstats jsonstats.cpp)
-  add_executable(jsonpointer jsonpointer.cpp)
-  add_executable(minify minify.cpp)
+  foreach(tool IN ITEMS json2json jsonstats jsonpointer minify)
+    add_executable("${tool}" "${tool}.cpp")
+    simdjson_apply_props("${tool}")
```
Comment (jkeiser) [inline tools/CMakeLists.txt:5]: Does this need to be a function somewhere, like add_cpp_test, or is this the only place we do it?

**Review by jkeiser — COMMENTED** [review]
I'm super happy we're getting cmake love :)

SIMDJSON_DEVELOPER_MODE makes perfect sense to me. The default should be just the stuff the user needs. And we don't have enough devs that flipping it would become a problem :)

For my edification, is there a user-defined place I can make it so that -DSIMDJSON_DEVELOPER_MODE=ON is always set for me? I blow the build away and re-run cmake a lot.

**Comment by friendlyanon** [comment]
The error in the `justlib-gcc10` workflow makes no sense.  
I don't know what to make of those ODR violations in `sanitize-clang10` and `sanitize-haswell-clang10`.  
The ondemand tests never finish compiling with 100% I/O usage on my end, so I can't really check those locally.

**Comment by friendlyanon** [comment]
> For my edification, is there a user-defined place I can make it so that -DSIMDJSON_DEVELOPER_MODE=ON is always set for me? I blow the build away and re-run cmake a lot.

Install CMake 3.19 on your machine and use [CMake Presets](https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html).

**Comment by jkeiser** [comment]
@friendlyanon PR #1514 is an attempt to deal with that. GCC sanitize takes a long time and I'm guessing the cause of the interruption is cmake timeouts.

File: .github/workflows/power-fuzz.yml [hunk 9]
```diff
@@ -32,7 +32,12 @@ jobs:
           run: |
             export CLANGSUFFIX="-7"
             apt-get -qq update
-            apt-get install -q -y clang-7 libfuzzer-7-dev cmake git wget zip ninja-build
+            apt-get install -q -y clang-7 libfuzzer-7-dev git wget zip ninja-build gnupg software-properties-common
+            wget -q -O - "https://raw.githubusercontent.com/simdjson/debian-ppa/master/key.gpg" | apt-key add -
+            apt-add-repository "deb https://raw.githubusercontent.com/simdjson/debian-ppa/master simdjson main"
+            apt-get -qq update
+            apt-get purge cmake cmake-data
+            apt-get -t simdjson -y install cmake
```
[the remark this answers is not in the document]
Comment (friendlyanon) [reply .github/workflows/power-fuzz.yml:40]: The documentation is seriously lacking wrt this flag, but to my understanding you can pin your requested versions to a specific `Suite` field in the debian repo's `Packages` metadata, which I defined to be `simdjson` in [our repo](https://github.com/simdjson/debian-ppa/blob/master/update.sh#L16).

I can't really say more and I'm not sure what I said is entirely correct either. The Debian/APT folks could take a look at their documentation and fix things up.

File: HACKING.md [hunk 10]
```diff
@@ -74,15 +86,15 @@ Other important files and directories:
   ```bash
   mkdir build
   cd build
-  cmake ..
+  cmake .. -D SIMDJSON_DEVELOPER_MODE=ON
```
[the remark this answers is not in the document]
Comment (friendlyanon) [reply HACKING.md:89]: CMake doesn't really care what comes first, unless it is one of those mode arguments like `--build` and `--install`, or `-P`.

I would be lying if I didn't admit that I tried to play things around in a way that I don't have to touch CI files *too much* due to the sheer number of them.

File: benchmark/checkperf.cmake (hunk 1 above)
Comment (friendlyanon) [reply benchmark/checkperf.cmake:63]: https://github.com/simdjson/simdjson/pull/1512/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR51-R53

The `env` shouldn't really be used here anyway, the variables could be passed as is via `-D` arguments, but in this PR my focus was clearly separating consumer and developer side of things.

File: cmake/developer-options.cmake (hunk 2 above)
Comment (friendlyanon) [reply cmake/developer-options.cmake:1]: More to come in later PRs!

**Comment by jkeiser** [comment]
@friendlyanon the ODR violations are because you changed those tests to link with `simdjson` instead of `simdjson-include-source`, but they do `#include "simdjson.cpp"` internally (because they are whiteboxing). That means things get defined twice on link.

File: cmake/simdjson-config.cmake.in (hunk 3 above)
Comment (friendlyanon) [reply cmake/simdjson-config.cmake.in:5]: I'm not sure what good is there in forcing it. I could use some explanation behind the logic here.

File: cmake/simdjson-props.cmake (hunk 4 above)
Comment (friendlyanon) [reply cmake/simdjson-props.cmake:1]: Since flags were previously accumulated via targets, those were causing issues with implementation details leaking into the install interface.

Now these commands will accumulate command calls into a string variable, then eventually flush them into a CMake script that is called with `include`.

Basically creates a file like this:
```cmake
target_compile_definitions("${target}" ${public} [==[arg1]==] [==[arg2]==])
target_compile_options("${target}" ${private} [==[arg1]==])
```

File: cmake/simdjson-user-cmakecache.cmake (hunk 5 above)
Comment (friendlyanon) [reply cmake/simdjson-user-cmakecache.cmake:2]: I turned this into a script that is called with `-P` mode, it's the first custom target after the developer mode delimiter in the root CML.

File: src/CMakeLists.txt (hunk 6 above)
Comment (friendlyanon) [reply src/CMakeLists.txt:1]: It was never a requirement and keeping things in as few files as possible could help with readability. Furthermore, the library definition itself I think should go in the root to be visible.

File: tests/CMakeLists.txt (hunk 7 above)
Comment (friendlyanon) [reply tests/CMakeLists.txt:12]: Maybe that's where the ODR violations come from in 2 of the CircleCI workflows? Things are a messy and it's very hard to understand dependencies between the targets.

File: tools/CMakeLists.txt (hunk 8 above)
Comment (friendlyanon) [reply tools/CMakeLists.txt:5]: Idon't think so, this is done only in this one place and only to make sure `simdjson_apply_props` applies link flags before anything else to satisfy ASan's needs.

File: .github/workflows/power-fuzz.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (jkeiser) [reply .github/workflows/power-fuzz.yml:40]: OK :) What does pinning the versions to the suite field do?

File: HACKING.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (jkeiser) [reply HACKING.md:89]: That's a perfectly valid reason. A note, though: I went into VS and did ctrl+shift+F with a regex search:

`cmake (.+) \.\.` -> `cmake .. $1`

And it seemed to touch just the right stuff. I don't have time to test it out right now, though, so YMMV.

File: cmake/simdjson-config.cmake.in (hunk 3 above)
Comment (jkeiser) [reply cmake/simdjson-config.cmake.in:5]: @lemire knows a lot more than I do :)

File: .github/workflows/power-fuzz.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (friendlyanon) [reply .github/workflows/power-fuzz.yml:40]: Makes sure our repo is picked over everything else. IIRC sets its priority to 900.

File: cmake/simdjson-props.cmake (hunk 4 above)
Comment (jkeiser) [reply cmake/simdjson-props.cmake:1]: Oh! So the idea is not to put these flags onto people who include simdjson? Hmm. It's a really good idea to separate them. I know a number of the flags are added on purpose, because they are needed when the user includes simdjson.h. Which ones should get removed?

Would a target we only use in simdjson (simdjson-internal or something) be a more idiomatic choice?

File: tests/CMakeLists.txt (hunk 7 above)
Comment (jkeiser) [reply tests/CMakeLists.txt:12]: Ah. I think it did.

**Comment by jkeiser** [comment]
I don't know what's up with the `justlib-gcc10` errors, but they are likely a result of your changes. One place to look is in the place where checkperf checks out master: it passes those variables, IIRC.

File: cmake/simdjson-props.cmake (hunk 4 above)
Comment (jkeiser) [reply cmake/simdjson-props.cmake:1]: @friendlyanon I think this is what `simdjson-internal-flags` was for.

The new approach does seem like a lot of extra code / concepts. But maybe the way we were doing it was abusing libraries, so that might be fine :)

File: benchmark/checkperf.cmake (hunk 1 above)
Comment (friendlyanon) [reply benchmark/checkperf.cmake]: I'm going to unresolve this, because I was dumb when I wrote the code. `CI` is only defined in the CI and this target is behind the developer mode switch and operates on developer mode targets, so setting the developer mode flag unconditionally makes the most sense.

**Comment by friendlyanon** [comment]
Okay, those ondemand tests take a heavy toll on the CI machines, the linker even started [spewing garbage nonstop](https://travis-ci.org/github/simdjson/simdjson/jobs/764135412).

I say that CircleCI job can be safely ignored. I don't even know what's the purpose of fetching the master branch and then building something from it.

**Comment by lemire** [comment]
>  I don't even know what's the purpose of fetching the master branch and then building something from it in the middle of the build.

It is a performance test, we compare the performance of the pull request with the main branch. It is relatively cheap as we only build a single binary.

**Comment by lemire** [comment]
The failing test is related to the following...

```yaml
  justlib-gcc10:
    description: Build just the library, install it and do a basic test
    executor: gcc10
    environment: { CMAKE_FLAGS: -DSIMDJSON_JUST_LIBRARY=ON }
    steps: [ cmake_build, cmake_install_test, cmake_installed_test_cxx20 ]
```

It should not be triggering the build of `parse` nor any checking out of the main branch?

**Comment by lemire** [comment]
@friendlyanon What is your expectation of what the following should do ?

```
cmake  -DSIMDJSON_JUST_LIBRARY=ON -B build && cmake --build build
```

**Comment by friendlyanon** [comment]
Good catch, I forgot to handle that variable.

To get only the library to be build in CI in the future, CMake has to be configured with `CI` set to a false value:
```cmake
CI= cmake # config args
```
or the `simdjson` target has to be built explicitly:
```cmake
cmake --build build --target simdjson
```

This is because the env variable `CI` set to a true value will force the configuration to happen in developer mode, so all targets are made available.

File: cmake/simdjson-props.cmake (hunk 4 above)
Comment (jkeiser) [reply cmake/simdjson-props.cmake:1]: @friendlyanon checking: did I understand the purpose of this correctly? Did `simdjson-internal-flags` not cover this use case before?

As a naive cmake user it seems more complicated than `simdjson-internal-flags` is all.

File: benchmark/checkperf.cmake (hunk 1 above)
Comment (jkeiser) [reply benchmark/checkperf.cmake]: How did this get resolved?

**Comment by jkeiser** [comment]
@lemire are you good with the JUST_LIBRARY fix? The only failing test appears to be `travis-ci`.

I think there are two unresolved things:

* [Threading logic](https://github.com/simdjson/simdjson/pull/1512#discussion_r598882693) (@lemire the thread one is for you, I think).
* [simdjson-props vs. simdjson-internal-flags](https://github.com/simdjson/simdjson/pull/1512#discussion_r598914367) -- I'll bend over backwards to accomodate the change if there's a good reason for it, but I can't figure out what makes the new logic better than `simdjson-internal-flags`, which had the benefit of not needing macros and such. (The answer might be "you are super abusing target_link_library by making targets that just provide flags." I am not a cmake expert :) )

**Comment by lemire** [comment]
@jkeiser I am not alarmed by the CMAKE_THREAD_PREFER_PTHREAD change.

**Comment by lemire** [comment]
@jkeiser The `SIMDJSON_JUST_LIBRARY` was always meant to be a hack so I do not care if it gets deprecated.

File: benchmark/checkperf.cmake (hunk 1 above)
Comment (friendlyanon) [reply benchmark/checkperf.cmake]: You marked the conversation as resolved.

If you mean in the code, then it's here https://github.com/simdjson/simdjson/commit/083c3fb10d7f5100e37ff0c6b13968d3f935443c

File: cmake/simdjson-props.cmake (hunk 4 above)
Comment (friendlyanon) [reply cmake/simdjson-props.cmake:1]: The problem with using targets for flags is that they will leak into the install interface.

Since there are tests, that essentially double the work by building the library again, I don't know if this could be removed. Maybe those tests could use some rework.

**Comment by friendlyanon** [comment]
So, what's the verdict on this?

**Comment by lemire** [comment]
@friendlyanon 



I think that @jkeiser expressed the following concern...


> I'll bend over backwards to accomodate the change if there's a good reason for it, but I can't figure out what makes the new logic better than simdjson-internal-flags, which had the benefit of not needing macros and such



It seems generally that we end up with more code and code that is relatively more complex. The goal of this PR, as I understand it, is to reduce the number of CMake targets.

I'd really like to understand better the benefits. I realize that you worked quite hard on this.

Would you consider breaking this down into multiple PRs, each with a clear rationale. For example, one of your motivations stated in the PR comment is to fix the static/dynamic lib builds... but this can be addressed by a much more limited set of changes...

https://github.com/simdjson/simdjson/pull/1544

I agree with the developer mode and this makes sense as well, but, again, it seems that it can be achieved using far fewer changes.

I understand that you wish to refactor the whole CMake build, but it is really hard for us to see whether this is easier to maintain. To me, it looks significantly more complicated and there is more code.

I am trying hard to live by the following approach:

"We do not make changes to simdjson without clearly identifiable benefits, which typically means either performance improvements, bug fixes or new features. Avoid bike-shedding: we all have opinions about how to write code, but we want to focus on what makes simdjson objectively better."

So I discourage refactoring in general, unless it can clearly enable new features, or more speed.

**Comment by lemire** [comment]
@friendlyanon I do realize that you are highly qualified and that you worked hard on this.

My concern is that there are hundreds of changes and it is just not possible for me to see what the consequences of all these changes are. I can appreciate the benefits of both the developer mode and the static/dynamic library change, but these are benefits that we can get with far smaller changes. So to me, it is really about making the changes in a way where we can say... "oh, ok, we are changing these lines, and here is why".

**Comment by friendlyanon** [comment]
The problem with the approach in #1544 is that it is duplicating work that CMake already does. `BUILD_SHARED_LIBS` is the canonical way to control the library types.

This PR would also be a shorter and simpler, if there were not tests that duplicate work by rebuilding the library again. I think there could be a better way to solve this issue, it needs a closer look.

My goal with the developer mode is to define very clear boundaries, much like in the case of componentized install rules. There are targets and code that need to be run only for the developer(s) of a project, other people who just want to build it are usually far greater in number, so it makes sense that the developer mode is the exception to the rule, i.e. it shouldn't be the default.

Of course if this PR were to be merged there would still be rough edges to mull over, like the tests mentioned above, but if things are sorted out properly, then that would boost maintainability by relying a lot more on CMake provided functionality instead of reinventing the wheel. I think you can agree that as of now the build tooling is piles of hacks and I don't blame you, building libraries it not a trivial task, the breadth of knowledge required is vast and spreading yourself thin can result in such things.

Most of the changes are really only just moving existing code around with minimal changes (so the diffs would still be big, regardless of how many PRs this is broken up into), so they are only active in developer mode, removing things from the install interface that shouldn't have been there to begin with (flags targets).  
The end goal would be to move flags and such out of the CMake scripts entirely and into presets (or toolchain files, if you don't want to require developers working on simdjson to use CMake 3.19+, but presets would definitely be better, as developers should always be on the latest not-point-oh release of CMake).

**Comment by lemire** [comment]
@friendlyanon 

I am game to try to merge this, and see how it goes...

About this...

>  This PR would also be a shorter and simpler, if there were not tests that duplicate work by rebuilding the library again. I think there could be a better way to solve this issue, it needs a closer look.

I am not sure what you refer to. I am pretty sure that the library artefact is only built once.

Of course, you need to check that you can build the single-header artefact (that's a test, not a build) and then you need to check for performance regressions, by comparing against master (but again, that's a test, not a build).

Importantly, we are not assuming that the consumers of this library are CMake users. People can either go with the single-header artefact or use CMake.

**Comment by lemire** [comment]
Ok. I took this PR out of a spin on my Apple M1 laptop (because it is a fast machine and it tends to produce reliable benchmarks).


I want to see how fast we can build. It should be the same speed, I guess, but let us check.

Let us try the parallelized build... If anyone cares a lot about speed, they will parallelize.

On our main branch... (I do it twice to get an estimate with error margin)

```
cmake --build buildtestpara --parallel  307.76s user 31.76s system 602% cpu 56.352 total
cmake --build buildtestpara2 --parallel  296.93s user 28.98s system 615% cpu 52.953 total
```

That is good, under a minute!!!

What about this PR?

```
cmake --build buildpara --parallel  247.07s user 26.17s system 585% cpu 46.629 total
cmake --build buildpara2 --parallel  247.92s user 25.88s system 596% cpu 45.930 total
```

Oh! It seems that this PR is consistently faster, at least for the parallelized builds on my Apple laptop.

**Comment by lemire** [comment]
Let us now run the tests. Again I am going to focus on parallelized tests.

Main branch... (I repeat three times starting from a new build)

```
ctest . --parallel  28.83s user 5.51s system 73% cpu 46.516 total
ctest . --parallel  13.84s user 4.49s system 93% cpu 19.611 total
ctest . --parallel  14.07s user 4.60s system 93% cpu 20.012 total
```

This PR...

```
ctest . --parallel  28.62s user 5.49s system 76% cpu 44.602 total
ctest . --parallel  12.51s user 3.83s system 93% cpu 17.481 total
ctest . --parallel  12.40s user 3.79s system 93% cpu 17.317 total
```


So it looks faster but there is a caveat. The old tests ran 74 tests, whereas this PR generates only 73 tests. Furthermore, on my machine,  checkperf fails systematically with the new PR.

**Comment by lemire** [comment]
In the main branch, I get...

```
74: The new code has a throughput in       3.20364 -- 3.21313
74: The reference code has a throughput in 3.20703 -- 3.2111
```

In the PR, I get...

```
73: The new code has a throughput in       3.18479 -- 3.18947
73: The reference code has a throughput in 3.20566 -- 3.2077
```

This seems like a small effect but it is puzzling that you'd fine a systematic performance degradation.

**Comment by lemire** [comment]
The `ondemand_tostring_tests` is missing from this PR. It would need to be added.

**Comment by lemire** [comment]
Regarding performance, one explanation is that the new build is static by default so we are not comparing like for like.

**Comment by lemire** [comment]
A static build on our main branch takes even longer...

```
cmake --build buildstatic --parallel  322.33s user 30.70s system 557% cpu 1:03.32 total
```

**Comment by lemire** [comment]
So it looks like static builds are slower in my configuration (that's odd but not entirely unexpected).

Here is checkperf with a static build on our main branch...
```
74: The new code has a throughput in       3.18411 -- 3.19082
74: The reference code has a throughput in 3.18211 -- 3.19015
```

Here is checkperf with the new branch...

```
1: The new code has a throughput in       3.18346 -- 3.19081
1: The reference code has a throughput in 3.20496 -- 3.2111
```

So what seems to be happening is that we don't compare like-for-like in the checkperf tests within the new branch. This may only be a temporary effect or it may be something that needs fixing. Let me read the build code again.

**Comment by lemire** [comment]
Ok. In  benchmark/checkperf.cmake, the static/dynamic preferences are not passed, but I assume that this will be resolved automatically in the future after we merge this.

File: tools/release.py [hunk 11]
```diff
@@ -145,11 +145,8 @@ def topaddedversionstring(major, minor, rev):
     sonumber += 1
 
 for line in fileinput.input(cmakefile, inplace=1, backup='.bak'):
-    line = re.sub('SIMDJSON_SEMANTIC_VERSION "\d+\.\d+\.\d+','SIMDJSON_SEMANTIC_VERSION "'+newversionstring, line.rstrip())
+    line = re.sub('    VERSION \d+\.\d+\.\d+','    VERSION '+newmajorversionstring+'.'+mewminorversionstring+'.'+newrevversionstring, line)
```
Comment (lemire) [inline tools/release.py:148]: You need line.rstrip() on this line because the line that is read includes the line ending control characters. Latter we call 'print' which inserts new  line ending control characters, hence effectively inserting new blank lines between each line. It is either  line.rstrip() or we pass an 'end' parameter to the 'print' function below.

File: cmake/simdjson-config.cmake.in (hunk 3 above)
Comment (lemire) [reply cmake/simdjson-config.cmake.in:5]: I do not think it matters much. The `CMAKE_THREAD_PREFER_PTHREAD` parameter is a workaround to avoid problems on some systems, but this can be fixed by the library user.

**Review by lemire — CHANGES_REQUESTED** [review]
Ok. Let us merge this. I am a bit nervous but @friendlyanon makes a good case.

The one blocking issue is that release.py script which will insert blank lines in CMakeLists.txt. Please fix that before we merge.

**Comment by lemire** [comment]
The missing `ondemand_tostring_tests` should get automatically added once we merge with our main branch. It is not a failure of this PR that the test is missing.

**Comment by lemire** [comment]
I am going to merge this and make the changes to the Python script in a later commit.

File: tools/release.py (hunk 11 above)
Comment (lemire) [reply tools/release.py:148]: Resolved in main branch.
